File: windows/PrismWin/PrismService.cs
using System.Net;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace PrismWin;

/// One neighbor returned by `/api/similarity`. Tiers share sha256/name/system and
/// carry whichever score applies (Jaccard, TLSH distance, cosine sim, shared count).
public sealed class SimilarNeighbor
{
    [JsonPropertyName("sha256")] public string Sha256 { get; set; } = "";
    [JsonPropertyName("name")] public string? Name { get; set; }
    [JsonPropertyName("system")] public string? System { get; set; }
    [JsonPropertyName("jaccard")] public double? Jaccard { get; set; }
    [JsonPropertyName("distance")] public double? Distance { get; set; }
    [JsonPropertyName("sim")] public double? Sim { get; set; }
    [JsonPropertyName("shared")] public long? Shared { get; set; }

    /// Human-readable score for the tier it came from.
    [JsonIgnore]
    public string? ScoreText =>
        Jaccard is { } j ? $"{j * 100:0}%"
        : Sim is { } s ? $"{s:0.00}"
        : Distance is { } d ? $"d={(int)d}"
        : Shared is { } n ? $"{n} shared"
        : null;
}

/// Response of `POST /api/similarity` (tiers default to empty when absent).
public sealed class SimilarityResponse
{
    [JsonPropertyName("identical_content")] public List<SimilarNeighbor> IdenticalContent { get; set; } = new();
    [JsonPropertyName("shared_files")] public List<SimilarNeighbor> SharedFiles { get; set; } = new();
    [JsonPropertyName("similar_chunks")] public List<SimilarNeighbor> SimilarChunks { get; set; } = new();
    [JsonPropertyName("exe_imports")] public List<SimilarNeighbor> ExeImports { get; set; } = new();
    [JsonPropertyName("exe_similar")] public List<SimilarNeighbor> ExeSimilar { get; set; } = new();
    [JsonPropertyName("audio_neighbors")] public List<SimilarNeighbor> AudioNeighbors { get; set; } = new();
    [JsonPropertyName("text_neighbors")] public List<SimilarNeighbor> TextNeighbors { get; set; } = new();

    /// (section title, neighbors) for non-empty tiers, in display order.
    [JsonIgnore]
    public List<(string Title, List<SimilarNeighbor> Neighbors)> Sections => new (string, List<SimilarNeighbor>)[]
    {
        ("Identical content", IdenticalContent),
        ("Shared files", SharedFiles),
        ("Similar chunks", SimilarChunks),
        ("Same boot imports", ExeImports),
        ("Similar executable", ExeSimilar),
        ("Shared audio tracks", AudioNeighbors),
        ("Semantically related (text)", TextNeighbors),
    }.Where(s => s.Item2.Count > 0).ToList();

    [JsonIgnore]
    public bool IsEmpty => Sections.Count == 0;
}

/// Thrown for non-2xx responses; carries the status code and body for the
/// chunked-upload resume (409) and rate-limit (429) handling.
public sealed class ServiceHttpException : Exception
{
    public int Code { get; }
    public string Body { get; }

    public ServiceHttpException(int code, string body)
        : base(code == 0 ? body.Trim() : $"Server error {code}: {body.Trim()}")
    {
        Code = code;
        Body = body;
    }
}

/// Client for the Prism web service. Base URL from `PRISM_WEB_URL` (default
/// `https://hiddenpalace.org`); optional `PRISM_MODERATION_TOKEN` auto-accepts
/// submits so they replace the live build instead of waiting in the queue.
public sealed class PrismService
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(120) };

    public Uri BaseUrl { get; }
    public string? ModerationToken { get; }

    public PrismService()
    {
        var raw = Environment.GetEnvironmentVariable("PRISM_WEB_URL");
        if (string.IsNullOrWhiteSpace(raw))
        {
            raw = "https://hiddenpalace.org";
        }
        BaseUrl = new Uri(raw.TrimEnd('/'));
        var token = Environment.GetEnvironmentVariable("PRISM_MODERATION_TOKEN");
        ModerationToken = string.IsNullOrEmpty(token) ? null : token;
    }

    public Uri BuildPage(string sha256) => new(BaseUrl, $"/build/{sha256}");

    /// POST the canonical record JSON to `/api/similarity`.
    public async Task<SimilarityResponse> FindSimilarAsync(string recordJson)
    {
        var body = await PerformAsync(HttpMethod.Post, "/api/similarity", recordJson, "application/json");
        return JsonSerializer.Deserialize<SimilarityResponse>(body) ?? new SimilarityResponse();
    }

    public sealed record SubmitResult(string Sha256, string Status);

    /// POST `{ nickname, record }` to `/api/submissions`.
    public async Task<SubmitResult> SubmitAsync(string recordJson, string nickname)
    {
        var payload = new JsonObject
        {
            ["nickname"] = nickname,
            ["record"] = JsonNode.Parse(recordJson),
        };
        var body = await PerformAsync(HttpMethod.Post, "/api/submissions", payload.ToJsonString(), "application/json");
        var v = JsonNode.Parse(body);
        return new SubmitResult(
            v?["sha256"]?.GetValue<string>() ?? "",
            v?["status"]?.GetValue<string>() ?? "queued");
    }

    /// Accept a queued submission with the moderation token, replacing the live build.
    public async Task AcceptAsync(string buildSha)
    {
        if (ModerationToken == null)
        {
            return;
        }
        using var req = new HttpRequestMessage(HttpMethod.Post, new Uri(BaseUrl, $"/api/submissions/{buildSha}"));
        req.Content = new StringContent("{\"action\":\"accept\"}", Encoding.UTF8, "application/json");
        req.Headers.Add("x-moderation-token", ModerationToken);
        await SendAsync(req);
    }

    /// GET `/api/submissions/<sha>/assets`: asset blobs the server still lacks.
    public async Task<List<string>> MissingAssetsAsync(string buildSha)
    {
        var body = await PerformAsync(HttpMethod.Get, $"/api/submissions/{buildSha}/assets", null, null);
        var v = JsonNode.Parse(body);
        return v?["missing"]?.AsArray().Select(n => n?.GetValue<string>())
            .Where(s => s != null).Select(s => s!).ToList() ?? new List<string>();
    }

    /// Upload chunk size, small enough to clear typical proxy body-size limits.
    private const int UploadChunkBytes = 4 * 1024 * 1024;

    /// Give up after this many consecutive rate-limit waits on one chunk.
    private const int MaxThrottleRetries = 30;

    /// PUT one asset blob under its build in resumable chunks: each request
    /// appends at `offset`, a 409 answers with the server's staged offset to
    /// resume from, and the final chunk returns `stored` (or `exists`).
    public async Task UploadAssetAsync(string buildSha, string assetSha, string filePath)
    {
        using var file = File.OpenRead(filePath);
        long offset = 0;
        var progress = new UploadProgress(file.Length);
        var throttled = 0;
        var buffer = new byte[UploadChunkBytes];
        while (true)
        {
            file.Seek(offset, SeekOrigin.Begin);
            var read = await file.ReadAsync(buffer.AsMemory(0, UploadChunkBytes));
            if (read == 0)
            {
                // The record claims more bytes than the local blob holds.
                throw new ServiceHttpException(0, "asset blob shorter than its record claims");
            }
            var url = new Uri(BaseUrl, $"/api/submissions/{buildSha}/assets/{assetSha}?offset={offset}");
            using var req = new HttpRequestMessage(HttpMethod.Put, url);
            req.Content = new ByteArrayContent(buffer, 0, read);
            req.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/octet-stream");
            try
            {
                var body = await SendAsync(req);
                var v = JsonNode.Parse(body);
                var status = v?["status"]?.GetValue<string>();
                if (status is "stored" or "exists")
                {
                    return;
                }
                if (status != "partial") throw new ServiceHttpException(0, "invalid upload status");
                offset = progress.Advance(TryField(body, "offset"));
                throttled = 0;
            }
            catch (ServiceHttpException e) when (e.Code == 409)
            {
                offset = progress.Resume(TryField(e.Body, "offset"));
            }
            catch (ServiceHttpException e) when (e.Code == 429)
            {
                // Rate limited. Wait out the window (the server's retryAfter
                // when present) and retry the same offset.
                throttled++;
                if (throttled > MaxThrottleRetries)
                {
                    throw;
                }
                var hint = TryDoubleField(e.Body, "retryAfter") ?? 5.0;
                await Task.Delay(TimeSpan.FromSeconds(Math.Clamp(hint, 1.0, 120.0)));
            }
        }
    }

    private static long? TryField(string json, string name)
    {
        try
        {
            return JsonNode.Parse(json)?[name]?.GetValue<long>();
        }
        catch
        {
            return null;
        }
    }

    private static double? TryDoubleField(string json, string name)
    {
        try
        {
            return JsonNode.Parse(json)?[name]?.GetValue<double>();
        }
        catch
        {
            return null;
        }
    }

    private async Task<string> PerformAsync(HttpMethod method, string path, string? body, string? contentType)
    {
        using var req = new HttpRequestMessage(method, new Uri(BaseUrl, path));
        if (body != null)
        {
            req.Content = new StringContent(body, Encoding.UTF8, contentType ?? "application/json");
        }
        return await SendAsync(req);
    }

    /// Run one request with the shared friendly-error mapping.
    private async Task<string> SendAsync(HttpRequestMessage req)
    {
        HttpResponseMessage resp;
        try
        {
            resp = await Http.SendAsync(req);
        }
        catch (TaskCanceledException)
        {
            throw new ServiceHttpException(0, $"Request to {BaseUrl.Host} timed out. Is the web service responding?");
        }
        catch (HttpRequestException e)
        {
            throw new ServiceHttpException(0, $"Cannot reach {BaseUrl.Host}: {e.Message}");
        }
        using (resp)
        {
            var body = await resp.Content.ReadAsStringAsync();
            if (!resp.IsSuccessStatusCode)
            {
                throw new ServiceHttpException((int)resp.StatusCode, body);
            }
            return body;
        }
    }
}

File: macos/Sources/PrismApp/PrismService.swift
import Foundation

/// One neighbor returned by `/api/similarity`. Tiers share sha256/name/system and
/// carry whichever score applies (Jaccard, TLSH distance, cosine sim, shared count).
struct SimilarNeighbor: Decodable, Identifiable {
    let sha256: String
    let name: String?
    let system: String?
    let jaccard: Double?
    let distance: Double?
    let sim: Double?
    let shared: Int?
    var id: String { sha256 }

    /// Human-readable score for the tier it came from.
    var scoreText: String? {
        if let jaccard { return String(format: "%.0f%%", jaccard * 100) }
        if let sim { return String(format: "%.2f", sim) }
        if let distance { return "d=\(Int(distance))" }
        if let shared { return "\(shared) shared" }
        return nil
    }
}

/// Response of `POST /api/similarity` (arrays default to empty when a tier is absent).
struct SimilarityResponse: Decodable {
    struct Query: Decodable { let sha256: String?; let name: String? }
    let query: Query
    let identicalContent: [SimilarNeighbor]
    let sharedFiles: [SimilarNeighbor]
    let similarChunks: [SimilarNeighbor]
    let exeImports: [SimilarNeighbor]
    let exeSimilar: [SimilarNeighbor]
    let audioNeighbors: [SimilarNeighbor]
    let textNeighbors: [SimilarNeighbor]

    private enum CodingKeys: String, CodingKey {
        case query
        case identicalContent = "identical_content"
        case sharedFiles = "shared_files"
        case similarChunks = "similar_chunks"
        case exeImports = "exe_imports"
        case exeSimilar = "exe_similar"
        case audioNeighbors = "audio_neighbors"
        case textNeighbors = "text_neighbors"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        query = try c.decode(Query.self, forKey: .query)
        func list(_ key: CodingKeys) -> [SimilarNeighbor] {
            (try? c.decode([SimilarNeighbor].self, forKey: key)) ?? []
        }
        identicalContent = list(.identicalContent)
        sharedFiles = list(.sharedFiles)
        similarChunks = list(.similarChunks)
        exeImports = list(.exeImports)
        exeSimilar = list(.exeSimilar)
        audioNeighbors = list(.audioNeighbors)
        textNeighbors = list(.textNeighbors)
    }

    /// (section title, neighbors) for non-empty tiers, in display order.
    var sections: [(String, [SimilarNeighbor])] {
        [
            ("Identical content", identicalContent),
            ("Shared files", sharedFiles),
            ("Similar chunks", similarChunks),
            ("Same boot imports", exeImports),
            ("Similar executable", exeSimilar),
            ("Shared audio tracks", audioNeighbors),
            ("Semantically related (text)", textNeighbors),
        ].filter { !$0.1.isEmpty }
    }

    var isEmpty: Bool { sections.isEmpty }
}

/// Read-only client for the Prism web service. Base URL from `PRISM_WEB_URL`
/// (default `https://hiddenpalace.org`; point it at a local dev server to test).
struct PrismService {
    /// The production default, built once without a force-unwrap.
    static let defaultBaseURL: URL = {
        var c = URLComponents()
        c.scheme = "https"
        c.host = "hiddenpalace.org"
        return c.url ?? URL(fileURLWithPath: "/")
    }()
    let baseURL: URL
    /// Optional moderation secret (`PRISM_MODERATION_TOKEN`). When set, a submit
    /// is auto-accepted so it replaces the live build instead of waiting in the queue.
    let moderationToken: String?

    init() {
        let raw = ProcessInfo.processInfo.environment["PRISM_WEB_URL"] ?? "https://hiddenpalace.org"
        baseURL = URL(string: raw) ?? PrismService.defaultBaseURL
        let token = ProcessInfo.processInfo.environment["PRISM_MODERATION_TOKEN"] ?? ""
        moderationToken = token.isEmpty ? nil : token
    }

    enum ServiceError: LocalizedError {
        case http(Int, String)
        case offline(String)
        case timedOut(String)
        case cancelled
        case transport(String, String)
        var errorDescription: String? {
            switch self {
            case let .http(code, msg): return "Server error \(code): \(msg)"
            case let .offline(msg): return "Cannot reach \(msg). Is the web service running?"
            case let .timedOut(msg): return "Request to \(msg) timed out. Is the web service responding?"
            case .cancelled: return "Request was cancelled."
            case let .transport(msg, detail): return "Network error talking to \(msg): \(detail)"
            }
        }
    }

    /// POST the canonical record JSON to `/api/similarity`.
    func findSimilar(recordJSON: String) async throws -> SimilarityResponse {
        let data = try await post(path: "/api/similarity", body: Data(recordJSON.utf8))
        return try JSONDecoder().decode(SimilarityResponse.self, from: data)
    }

    struct SubmitResult: Decodable { let sha256: String; let status: String }

    /// POST `{ nickname, record }` to `/api/submissions`.
    func submit(recordJSON: String, nickname: String) async throws -> SubmitResult {
        let record = try JSONSerialization.jsonObject(with: Data(recordJSON.utf8))
        let payload: [String: Any] = ["nickname": nickname, "record": record]
        let body = try JSONSerialization.data(withJSONObject: payload)
        let data = try await post(path: "/api/submissions", body: body)
        return try JSONDecoder().decode(SubmitResult.self, from: data)
    }

    /// Accept a queued submission with the moderation token, replacing the live build.
    func accept(buildSha: String) async throws {
        guard let token = moderationToken else { return }
        var req = URLRequest(url: baseURL.appendingPathComponent("/api/submissions/\(buildSha)"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(token, forHTTPHeaderField: "x-moderation-token")
        req.httpBody = Data(#"{"action":"accept"}"#.utf8)
        _ = try await perform(req)
    }

    private struct AssetsStatus: Decodable { let missing: [String] }

    /// GET `/api/submissions/<sha>/assets` — asset blobs the server still lacks.
    func missingAssets(buildSha: String) async throws -> [String] {
        var req = URLRequest(url: baseURL.appendingPathComponent("/api/submissions/\(buildSha)/assets"))
        req.httpMethod = "GET"
        let data = try await perform(req)
        return try JSONDecoder().decode(AssetsStatus.self, from: data).missing
    }

    /// Upload chunk size — small enough to clear typical proxy body-size limits.
    private static let uploadChunkBytes = 4 * 1024 * 1024

    /// One chunk response: `partial` carries the next offset; `stored`/`exists`
    /// end the upload. A 409 body carries only `offset` (the staged size); a
    /// 429 body carries `retryAfter` (seconds until the rate window resets).
    private struct ChunkResult: Decodable {
        let status: String?
        let offset: UInt64?
        let retryAfter: Double?
    }

    /// Give up after this many consecutive rate-limit waits on one chunk.
    private static let maxThrottleRetries = 30

    /// PUT one asset blob under its build in resumable chunks: each request
    /// appends at `offset`, a 409 answers with the server's staged offset to
    /// resume from, and the final chunk returns `stored` (or `exists`).
    func uploadAsset(buildSha: String, assetSha: String, fileURL: URL) async throws {
        let fh = try FileHandle(forReadingFrom: fileURL)
        defer { try? fh.close() }
        var offset: UInt64 = 0
        var progress = UploadProgress(size: try fh.seekToEnd())
        var throttled = 0
        while true {
            try fh.seek(toOffset: offset)
            let chunk = try fh.read(upToCount: Self.uploadChunkBytes) ?? Data()
            if chunk.isEmpty {
                // The record claims more bytes than the local blob holds.
                throw ServiceError.transport(baseURL.absoluteString, "asset blob shorter than its record claims")
            }
            var req = URLRequest(url: assetChunkURL(buildSha: buildSha, assetSha: assetSha, offset: offset))
            req.httpMethod = "PUT"
            req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
            req.httpBody = chunk
            do {
                let data = try await perform(req)
                let r = try JSONDecoder().decode(ChunkResult.self, from: data)
                if r.status == "stored" || r.status == "exists" { return }
                guard r.status == "partial" else { throw UploadProgress.Invalid.progress }
                offset = try progress.advance(r.offset)
                throttled = 0
            } catch let ServiceError.http(code, body) where code == 409 {
                let staged = (try? JSONDecoder().decode(ChunkResult.self, from: Data(body.utf8)))?.offset
                offset = try progress.resume(staged)
            } catch let ServiceError.http(code, body) where code == 429 {
                // Rate limited — wait out the window (the server's retryAfter
                // when present) and retry the same offset.
                throttled += 1
                if throttled > Self.maxThrottleRetries { throw ServiceError.http(code, body) }
                let hint = (try? JSONDecoder().decode(ChunkResult.self, from: Data(body.utf8)))?.retryAfter
                let seconds = min(max(hint ?? 5, 1), 120)
                try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            }
        }
    }

    private func assetChunkURL(buildSha: String, assetSha: String, offset: UInt64) -> URL {
        let url = baseURL.appendingPathComponent("/api/submissions/\(buildSha)/assets/\(assetSha)")
        var comps = URLComponents(url: url, resolvingAgainstBaseURL: false)
        comps?.queryItems = [URLQueryItem(name: "offset", value: String(offset))]
        return comps?.url ?? url
    }

    private func post(path: String, body: Data) async throws -> Data {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        return try await perform(req)
    }

    /// Run one request with the shared friendly-error mapping.
    private func perform(_ req: URLRequest) async throws -> Data {
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: req)
        } catch let urlError as URLError {
            let host = baseURL.absoluteString
            switch urlError.code {
            case .timedOut:
                throw ServiceError.timedOut(host)
            case .cancelled:
                throw ServiceError.cancelled
            case .notConnectedToInternet, .cannotConnectToHost, .cannotFindHost, .networkConnectionLost, .dnsLookupFailed:
                throw ServiceError.offline(host)
            default:
                throw ServiceError.transport(host, urlError.localizedDescription)
            }
        } catch {
            throw ServiceError.transport(baseURL.absoluteString, error.localizedDescription)
        }
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(code) else {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw ServiceError.http(code, msg)
        }
        return data
    }
}
